Implemented Resource categories, subcategories, and item list screens

Derek-Chu opened the pull request.

Derek-Chu committed:
Add categories fetch and buttons

Derek-Chu committed:
Begin implementation of subcategories

Derek-Chu committed:
Implemented Subcategories and refactored Categories

Added new styles. Added more loading states to give more feedback to user upon pressing button. Disabled buttons/TouchableOpacities upon click event. Added variables.js for common strings.

Derek-Chu committed:
Implement ResourcesItemListScreen

Derek-Chu committed:
Reduce clutter in style.js

vfcheung commented on src/screens/Resources/HomeScreen.js:
```diff
@@ -0,0 +1,167 @@
+      .where('parent', '==', id)
+      .get()
+      .then((subcategorySnapshot) => {
+        if (!subcategorySnapshot.empty) {
+          navigation.push(nav.subcategories, { snapshot: subcategorySnapshot, header: title });
+        } else {
+          firestore()
+            .collection(collections.items)
+            .where('parent', '==', id)
+            .get()
+            .then((itemSnapshot) => {
+              navigation.push(nav.itemList, { snapshot: itemSnapshot, header: title });
+            })
+            .catch((error) => {
+              setErrorMessage(error.message);
+              setLoading(false);
+            });
+        }
```
So this makes sense to me because we discussed how to design this feature, but someone new to this code might find it hard to parse these lines. Can we refactor? In fact `async` and `await` might help to make it easier to parse.

vfcheung commented on src/screens/Resources/HomeScreen.js:
```diff
@@ -0,0 +1,167 @@
+              >
+                <Image source={{ uri: thumbnail.src }} style={HomeStyles.categoryImage} />
+                <Text style={HomeStyles.categoryText}>
+                  {`${title}`}
```
You can just put `{title}`

vfcheung commented on src/screens/Resources/variables.js:
```diff
@@ -0,0 +1,11 @@
+export const collections = {
```
We can probably rename this file to `constants` and move it up so other components can use it.

vfcheung commented on src/screens/Resources/HomeScreen.js:
```diff
@@ -0,0 +1,167 @@
+        && (
+        <ActivityIndicator size="large" />
+        )}
+      <View style={HomeStyles.categoryButtonView}>
+        { !initialLoad
+          && (snapshot.docs.map((doc) => {
+            const {
+              title, thumbnail,
+            } = doc.data();
+
+            return (
+              <TouchableOpacity
+                style={HomeStyles.categoryButton}
+                key={doc.id}
+                disabled={loading}
+                onPress={() => loadScreen(doc.id, title)}
+              >
+                <Image source={{ uri: thumbnail.src }} style={HomeStyles.categoryImage} />
+                <Text style={HomeStyles.categoryText}>
+                  {`${title}`}
+                </Text>
+              </TouchableOpacity>
+            );
+          }))}
+      </View>
```
Might be worthwhile to create a separate component for this list so this component is less monolithic.

vfcheung commented on src/screens/Resources/ResourcesItemListScreen.js:
```diff
@@ -0,0 +1,183 @@
+            }}
+          >
+            More
+            {' '}
```
Use `&nbsp` (non-breaking space)

vfcheung commented on src/screens/Resources/ResourcesItemListScreen.js:
```diff
@@ -0,0 +1,183 @@
+
+  function singleOrMultipleDocs() {
+    // returns a collapsible if there is only 1 doc
+    if (snapshot.docs.length === 1) {
```
What if snapshot is empty?

Derek-Chu replied:
if snapshot is empty it works fine, there's just no interactive parts to the screen.

vfcheung commented on src/screens/Resources/ResourcesItemListScreen.js:
```diff
@@ -0,0 +1,183 @@
+      <View style={ItemListStyles.accordionHeader}>
+        <View style={ItemListStyles.accordionHeaderText}>
+          <Text style={ItemListStyles.resourceText}>
+            {doc.data().title}
```
In case we have to do null checks, calling `doc.data()` here might make that hard to do. Can we call `doc.data()` in one place, store the result pass it to these components?

vfcheung commented on src/screens/Resources/ResourcesItemListScreen.js:
```diff
@@ -0,0 +1,183 @@
+    if (snapshot.docs.length === 1) {
+      const doc = snapshot.docs[0];
+      return (
+        <View>
```
These components can be refactored into smaller components.

vfcheung commented on src/screens/Resources/ResourcesItemListScreen.js:
```diff
@@ -0,0 +1,183 @@
+
+  function renderContent(doc) {
+    return (
+      <View style={ItemListStyles.accordionContent}>
```
This can be refactored into its own component

vfcheung commented on src/screens/Resources/ResourcesSubcategoriesScreen.js:
```diff
@@ -0,0 +1,120 @@
+  const [loading, setLoading] = useState(false);
+  const [errorMessage, setErrorMessage] = useState(null);
+
+  const loadScreen = useCallback((id, title) => {
```
Can we refactor this with the other `loadScreen` into one file?

vfcheung commented on src/screens/Resources/ResourcesSubcategoriesScreen.js:
```diff
@@ -0,0 +1,120 @@
+        && (
+        <ActivityIndicator size="large" />
+        )}
+      { snapshot.docs.map((doc) => {
```
Can make this its own component

vfcheung reviewed (changes requested):
Good work, let's work on these changes for now.

Derek-Chu committed:
Refactor Home and Subcategories Screens

Moved variables.js to constants folder with a separate file for each variable. Made a LoadScreen.js for both HomeScreen and ResourcesSubcategoriesScreen to use. Made separate files for creating Home and Subcategory icons.

Derek-Chu committed:
Refactor ResourcesItemListScreen

Derek-Chu committed:
Fix merge conflicts

Derek-Chu committed:
Fix more merge conflicts

enmingliu commented on android/app/src/main/res/raw/app.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "name": "PTVMobileApp",
+  "displayName": "PTVMobileApp"
+}
```
Github is warning me about there not being a newline at the end of this file, not sure if you intended to not leave a newline here.

enmingliu commented on src/screens/Resources/HomeScreen.js:
```diff
@@ -0,0 +1,107 @@
+  const [loadingNextScreen, err, loadScreen] = useLoadScreen(navigation);
+  const [errorMessage, setErrorMessage] = useState(err);
+  const [snapshot, setSnapshot] = useState(null);
+  const [initialLoad, setInitialLoad] = useState(true);
```
Maybe its better to set this state to false at the start and set it to true when we actually query from firestore and have it represent just `loading` in case we have something in the future that reloads the props apart from just opening the app.

enmingliu commented on src/screens/Resources/LoadScreen.js:
```diff
@@ -0,0 +1,46 @@
+      .get();
+  }
+
+  async function loadScreen(id, title) {
```
I really like the idea of having a function to handle querying & loading into arbitrary subcategories/subcategory lists, but is it possible for us to be in a loading state in the newly pushed screen (ie., showing nothing but an activity indicator) instead of waiting for the query to complete in the old screen? I ask this because having the activity indicator shifting ui elements around in the old screen while locking any ui input isn't very natural, especially if the query fails for some reason. (vs being in a new screen and showing an error message there)

Derek-Chu replied:
to my knowledge not really. the reason why we handled it here is because we don't know whether we should be routing to a subcategories screen or a resources item list screen. because of that we need to query firebase to see which one we need to go to. and since we query firebase, i took the liberty of just extracting the snapshot and passing it to the next screen anyways. i'm open to any suggestions for how to change this tho!

enmingliu replied:
I think we could have a staging screen whose only purpose is to indicate that its in the process of loading the subcategories/resource item list, like how the facebook app shows grey boxes before all the content loads, since I think the issue is mainly with having the activity indicator & error message on the previous screen.

Derek-Chu replied:
hmm i see and then display the error msg and the activity indicator in this staging screen?

enmingliu replied:
Yup, exactly

enmingliu commented on src/screens/Resources/LoadScreen.js:
```diff
@@ -0,0 +1,46 @@
+  async function loadScreen(id, title) {
+    setLoading(true);
+
+    try {
```
Is there a specific reason we're using a try/catch block here instead of a promise off of firestore().collection().get()?

Derek-Chu replied:
we do that because otherwise it gets kinda messy and unclear bc i have to keep changing .then() statements. the async await generally makes it a bit clearer.

enmingliu replied:
Sounds good!

enmingliu commented on src/screens/Resources/ResourcesComponents/SingleItemList.js:
```diff
@@ -0,0 +1,96 @@
+  );
+}
+
+const SingleItemStyles = StyleSheet.create({
```
I'm not sure exactly what causes this problem, but the left/right margins for each list item are different when the accordion is extended vs not. (at least on iOS simulator, transportation -> big red bus) Is this intended?

Derek-Chu replied:
interesting but i cant test since i dont have an ios environment :(

enmingliu replied:
Ok, I don't think this is too big a deal, I can probably take another look at it after we merge this pr.

enmingliu commented on src/screens/Resources/ResourcesComponents/MultipleItemList.js:
```diff
@@ -0,0 +1,108 @@
+  );
+}
+
+const MultipleItemStyles = StyleSheet.create({
```
Also, it feels like a lot of styles are shared between MultipleItemStyles and SingleItemStyles, it would be great if we could reduce code reuse.

Derek-Chu replied:
I'll do this for now but I think the styles can be looked at more thoroughly in a future pr as well since i think this one is mostly a functionality-based one

enm1 commented:
This is a huge change, but it looks really good! Great work!

Derek-Chu committed:
Refactor styles and loading indicators
